python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,340 @@
+            self.connection.close()
+
+    def enable_receiving(self, loop=None):
+        '''Schedules the receive loop to run on the given loop.'''
+        self.receive_task = asyncio.ensure_future(self._receive_loop(),
+                                                  loop=loop)
+
+    @classmethod
+    async def from_socket(cls, sockname=None):
+        '''Create a new AIOClient using Unix transport and BSER Codec
+        connecting to the specified socket. If the specified socket is None,
+        then resolve the socket path automatically.
+
+        This method also schedules the receive loop to run on the event loop.
+
+        This method is a coroutine.'''
+        if not sockname:
+            sockname = await _resolve_sockname()
+        transport = AsyncUnixSocketTransport()
+        await transport.activate(sockname=sockname)
+        connection = AsyncBserCodec(transport)
+        obj = cls(connection)
+        obj.enable_receiving()
+        return obj
+
+    async def _receive_loop(self):
+        '''Receive the response to a request made to the Watchman service.
+
+        Note that when trying to receive a PDU from the Watchman service,
+        we might get a unilateral response to a subscription or log, so these
+        are processed and queued up for later retrieval. This function only
+        returns when a non-unilateral response is received.'''
+
+        while True:
+            response = await self.connection.receive()
+            if self._is_unilateral(response):
+                await self._process_unilateral_response(response)
+            else:
+                await self.bilateral_response_queue.put(response)
+
+    def _check_error(self, res):
+        if 'error' in res:
+            raise CommandError(res['error'])
+
+    def _check_recieve_loop(self):
+        if self.receive_task is None:
+            raise ReceiveLoopError('Recieve loop was not started.')
+
+    def _is_unilateral(self, res):
+        return 'unilateral' in res and res['unilateral'] or \
+               'subscription' in res or \
+               'log' in res
+
+    def _ensure_subscription_queue_exists(self, name, root):
+        # Note this function must be called from an async function on only one
+        # event loop.
+        if not root in self.sub_by_root:
+            self.sub_by_root[root] = {}
+        if not name in self.sub_by_root[root]:
+            self.sub_by_root[root][name] = asyncio.Queue()
+
+    async def _process_unilateral_response(self, response):
+        if 'log' in response:
+            await self.logs.put(response['log'])
+
+        elif 'subscription' in response:
+            sub = response['subscription']
+            root = os.path.normcase(response['root'])
+            self._ensure_subscription_queue_exists(sub, root)
+            await self.sub_by_root[root][sub].put(response)
+
+        elif self._is_unilateral(response):
+            raise WatchmanError('Unknown unilateral response.')
```

Comment:
You could add the keys in the response to the error for the user to be able to identify what that response contained. Or the entire response, actually, it might be helpful for debugging.